### DISCUSSION TITLE
Long-running Job Execution States

### DISCUSSION THREAD

**Discussion by davidhjones** [quartznet/quartznet#919] [General] [answered]
I'm looking to solve a problem, and I'm wondering if I could customize quartz to fit my needs. Let me explain the problem.

I have tasks which are scheduled on a recurring basis (e.g. Fire now, and every 4 hours afterwards). These jobs can take a lot of time (average 2 hours, up to 12 hours). The system that is executing the jobs is Azure Batch. I need a system which will monitor on-going tasks while they're executing in Azure Batch, such that jobs of the same "type" will not be executed while the same "type" job is currently executing. The way that it works is I queue a Task in Azure Batch, which will update me when it has completed. Currently I've been able to use Quartz to queue the Task in Azure Batch. 

While Quartz does an amazing job at scheduling the work, and triggering to execute the Job, I haven't found a way to manage long-running tasks. Obviously I could maintain a monitor within the _Execute_ implementation, but I'd be afraid of performance constraints for holding onto the process while the task is executing. Ideally I'd like to implement a more intelligent Job which can maintain States (e.g. "Queued", "Started", "Executing", "Finalizing", "Finished") -- basically I'd like to setup an async framework such that I can call `job.Finish()` (For example), and at that point Quartz would forget about this Job (as it does at the end of Execute). 

Thoughts?

**Comment by lahma** [comment, marked as answer]
I personally am not very familiar with Azure Batch (yet). In this case Azure Batch is basically what Quartz calls JobRunShell (the container handling Execute), but that would be probably quite difficult to model as it's out-of-process. 

I'm just throwing ideas here, but what about having a trigger on Quartz's side that handles the start and records state "Started", then re-schedules itself to run after say 5 minutes and on each invocation checks current status and updates the trigger's data map. On each re-schedule run would check batch status until finds it finished, no longer needing to re-schedule itself. This requires Quartz to run against persistent store to ensure data is not lost.

To create complete history and state I would advice having separate DB model for your batch jobs that store the current state of each Azure batch. This might also mean that you don't need to use trigger job data map, you could just query you own table on each run and the re-schedule as needed.

Each job (job key: group + name) obeys the `[DisallowConcurrentExecution]` attribute. So you would need to have a job created per Azure Batch job your want to protect (same job type is fine, as long as it has different job key). But with the earlier idea this doesn't work as expected when you trigger it just to check status and what to do.

So basically this means handling job/trigger state and checking whether can schedule the actual Azure Batch job or not (is the same key already running or not). So something with custom logic and state persistence could work. 

In ideal world Quartz would support such remote agents/schedulers like Azure Batch, but we are not there yet 🙂

**Reply by davidhjones** [reply]
Thanks for the prompt reply @lahma. Interesting approach to the problem. I do wonder if maintaining the execution status within the job data map / rescheduling will become cumbersome. 

As an alternative, I'm thinking about synergizing the execution Context of a Job with the Azure-Batch run -- I.E. running the Azure Batch task within `IJob.Execute` and awaiting its completion (`await Task.Delay` -- check status periodically, or setup Event-driven solution to "notify" job of its completion). Would you have any concerns about thread pool exhaustion or other issues with an approach that wouldn't finish `IJob.Execute` until the Azure Batch job had completed?

**Reply by lahma** [reply]
Well I think this boils down to idempotency (to jump into the computer science wagon). You need to be sure you can re-run your job without problems. If there's a timeout or computer restarts, process crash, whatever, you need to be sure you can track whether it's OK to start over. So always be prepared (maybe via external status storage) to check what's the status (started, running, completed).

You can create a long running job in Quartz for sure. You will deplete the thread pool based on allowed concurrent jobs. So you can configure any number of jobs allowed, especially in this case where jobs are waiting - not CPU bound. You can have 100 000 allowed concurrent jobs (if they don't kill the server), but be aware if something crashes they can be re-run and cause side-effects if you haven't planned for idempotency.
 
So you need some way of keeping track of status, whether it's job data map or external storage, re-fires/retries may happen.
